apis/core/v1beta1/openstackdataplane_types.go

```diff
@@ -0,0 +1,109 @@
+/*
+Copyright 2023.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package v1beta1
+
+import (
+	dataplanev1beta1 "github.com/openstack-k8s-operators/dataplane-operator/api/v1beta1"
+	condition "github.com/openstack-k8s-operators/lib-common/modules/common/condition"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// EDIT THIS FILE!  THIS IS SCAFFOLDING FOR YOU TO OWN!
+// NOTE: json tags are required.  Any new fields you add must have json tags for the fields to be serialized.
+
+// OpenStackDataPlaneSpec defines the desired state of OpenStackDataPlane
+type OpenStackDataPlaneSpec struct {
+
+	// +kubebuilder:validation:Optional
+	// Nodes - List of nodes
+	Nodes map[string]dataplanev1beta1.OpenStackDataPlaneNodeSpec `json:"nodes,omitempty"`
+	// +kubebuilder:validation:Optional
+	// Roles - List of roles
+	Roles map[string]dataplanev1beta1.OpenStackDataPlaneRoleSpec `json:"roles,omitempty"`
```

Comment:
Maintaining the list of nodes and roles here wasn't exactly how I was thinking this would work, however I'm open to it. I thought we would just label the roles with the dataplane they are in like we do for nodes in a role.

With how you have it defined here, we could create the DataPlane all in one YAML doc, which might make for a nicer UX interface. Alternatively, you need to create the DataPlane, Role, and Nodes separately and then trigger a deploy.

Another thought I've had is to associate these items with other top level CRDs, DataPlaneBinding that would group roles in a dataplane, and DataPlaneRoleBinding that would group nodes in a role.

Either way, we need the DeployStrategy field here as well so we can trigger deploys on the DataPlane CR.

Reply:
DataPlaneRoleBinding sounds more idiomatic to the kubernetes world, and the label approach is more consistent with the code we already have